---
output: github_document
---

# crispR

<!-- badges: start -->
<!-- badges: end -->

I wrote the functions directly in a `R package`. This facilitates the installation of all the dependencies...

## Installation

``` r
# install.packages("devtools")
devtools::install_github("c1au6i0/crispR")
```
## Part1 Answers

### a) the code for the function

You can access the code of the function `find_proto` ([here](https://github.com/c1au6i0/crispR/blob/master/R/protospacers.R)).

### b) the code to call the function with the example variables (and others, if desired) 

```{r}
library(crispR)
find_proto(d_seq = "TGATCTACTAGAGACTACTAACGGGGATACATAG",
           l = 2,
           PAM = "NGG")
```

..or using DNA of the Dopamine Transporter (`DAT` internal data):

```{r DAT}
library(crispR)
print(DAT)
```

```{r example2}
library(crispR)
find_proto(d_seq = DAT, 
           l = 20, 
           PAM = "NGG")
```

### c) the time complexity for the function (in big-O notation)

I am not explicitly using any loop, but  my function is in any case iterating and looking at each nucleotide of the sequence by using `grep` (`stringr` and regular expressions). 

time Complexity: *O(n)*

## Part2 Answers.

### a)  The code for the function

You can access the code of the function `find_FASTA` ([here](https://github.com/c1au6i0/crispR/blob/master/R/protospacers.R)).

###	b)  The source of the FASTA file used for the reference genome in the example problem 

I downloaded the Reference Genome Sequence GRCh38 from [here](https://www.ncbi.nlm.nih.gov/genome/guide/human/).

### c)  How many candidate guide (protospacer) sequences were identified in the example problem 

A total of 54 protospacers were identified on strand (+). Please note the arguments "start", "end" and "l" are 1-indexed and intervals are fully closed.

### d)  The list of candidate guide (protospacer) sequences in a tab-delimited file...

A tab-delimited file can be downloaded [here](https://1drv.ms/t/s!Am3aUTxhPMS8iZoyhzb0kQcSU3ITOw?e=BL0aqe).

### Dependencies

All the dependencies are listed in the `Description` file in my github account [here](https://github.com/c1au6i0/crispR/blob/master/DESCRIPTION).







